magnolia/shared/src/main/scala-dotty/zio/config/magnolia/DeriveConfig.scala

```diff
@@ -273,23 +309,37 @@ object DeriveConfig {
     else
       val listOfDesc =
         fieldNames.zip(allDescs).map { case (fieldName, desc) =>
-          val fieldDesc = tryAllKeys(desc.desc, Some(fieldName.originalName), fieldName.alternativeNames)
+          val fieldDesc = tryAllKeys(desc.desc, Some(fieldName.originalName), fieldName.alternativeNames, keyModifiers)
           fieldName.descriptions.foldRight(fieldDesc)((doc, desc) => desc ?? doc)
         }
 
       val descOfList =
-        Config.collectAll(listOfDesc.head, listOfDesc.tail: _*)
+        Config.collectAll(listOfDesc.head, listOfDesc.tail*)
 
-      DeriveConfig(descOfList.map(f), Some(Metadata.Product(productName, fieldNames)))
+      DeriveConfig(descOfList.map(f), Some(Metadata.Product(productName, fieldNames, keyModifiers)))
 
   def tryAllKeys[A](
     desc: Config[A],
     originalKey: Option[String],
-    alternativeKeys: List[String]
+    alternativeKeys: List[String],
+    keyModifiers: List[KeyModifier]
   ): Config[A] =
+
+    val sortedKeyModifiers = keyModifiers.sortWith {
+      case (a: CaseModifier, b: CaseModifier) => false
+      case (a: CaseModifier, _)               => false
+      case (_, b: CaseModifier)               => true
+      case _                                  => false
+    }
+
+    val modifyKey: String => String =
+      sortedKeyModifiers.map(KeyModifier.getModifierFunction).foldLeft(_)((key, modifier) => modifier(key))
+
     alternativeKeys match {
-      case Nil  => originalKey.fold(desc)(desc.nested(_))
-      case keys => keys.view.map(desc.nested(_)).reduce(_ orElse _)
+      case Nil  =>
+        originalKey.fold(desc)(k => desc.nested(modifyKey(k)))
+      // case keys => keys.view.map(k => desc.nested(modifyKey(k))).reduce(_ orElse _) // Looks like the Scala 3 implementation modifies alternative names while the Scala 2 implementations treats them as is.
+      case keys => keys.view.map(k => desc.nested(k)).reduce(_ orElse _)
```

Comment:
please review this carefully, because I think L341 should be enabled and L342 disabled. This changes the existing behaviour only to match the Scala 2 behaviour.

Reply:
L340 I already adjusted.